File: bindings/php/AnimationEffect.php
<?php
namespace Stencil;

final class AnimationEffect {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;
    private ?object $borrowedFrom;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned, ?object $borrowedFrom = null) {
        $this->ptr = $ptr;
        $this->owned = $owned;
        $this->borrowedFrom = $borrowedFrom;
    }

    public static function create( $duration_ms) {
        $ret = Lib::ffi()->AnimationEffect_create($duration_ms);
        return new AnimationEffect($ret, true);
    }

    public static function instant() {
        $ret = Lib::ffi()->AnimationEffect_instant();
        return new AnimationEffect($ret, true);
    }

    public static function popIn( $duration_ms) {
        $ret = Lib::ffi()->AnimationEffect_pop_in($duration_ms);
        return new AnimationEffect($ret, true);
    }

    public static function dropIn( $duration_ms,  $height) {
        $ret = Lib::ffi()->AnimationEffect_drop_in($duration_ms, $height);
        return new AnimationEffect($ret, true);
    }

    public static function dropAndPop( $duration_ms,  $height) {
        $ret = Lib::ffi()->AnimationEffect_drop_and_pop($duration_ms, $height);
        return new AnimationEffect($ret, true);
    }

    public static function spinIn( $duration_ms,  $turns) {
        $ret = Lib::ffi()->AnimationEffect_spin_in($duration_ms, $turns);
        return new AnimationEffect($ret, true);
    }

    public static function turntable( $duration_ms) {
        $ret = Lib::ffi()->AnimationEffect_turntable($duration_ms);
        return new AnimationEffect($ret, true);
    }

    public function addTween(string $property_name,  $from,  $to, string $easing_name) {
        $__n0 = strlen($property_name);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $property_name, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $__n3 = strlen($easing_name);
        $__view3 = Lib::ffi()->new('DiplomatStringView');
        if ($__n3 > 0) {
            $__buf3 = Lib::ffi()->new("uint8_t[" . $__n3 . "]", false);
            \FFI::memcpy($__buf3, $easing_name, $__n3);
            $__view3->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf3[0]));
        } else {
            $__view3->data = null;
        }
        $__view3->len = $__n3;
        $result = Lib::ffi()->AnimationEffect_add_tween($this->ptr, $__view0, $from, $to, $__view3);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function addKeyframe(string $property_name,  $at,  $value, string $easing_name) {
        $__n0 = strlen($property_name);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $property_name, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $__n3 = strlen($easing_name);
        $__view3 = Lib::ffi()->new('DiplomatStringView');
        if ($__n3 > 0) {
            $__buf3 = Lib::ffi()->new("uint8_t[" . $__n3 . "]", false);
            \FFI::memcpy($__buf3, $easing_name, $__n3);
            $__view3->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf3[0]));
        } else {
            $__view3->data = null;
        }
        $__view3->len = $__n3;
        $result = Lib::ffi()->AnimationEffect_add_keyframe($this->ptr, $__view0, $at, $value, $__view3);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function setRepeatForever() {
        Lib::ffi()->AnimationEffect_set_repeat_forever($this->ptr);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->AnimationEffect_destroy($this->ptr);
        }
    }
}

File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/AnimationEffect.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface AnimationEffectLib: Library {
    fun AnimationEffect_destroy(handle: Pointer)
    fun AnimationEffect_create(durationMs: Float): Pointer
    fun AnimationEffect_instant(): Pointer
    fun AnimationEffect_pop_in(durationMs: Float): Pointer
    fun AnimationEffect_drop_in(durationMs: Float, height: Float): Pointer
    fun AnimationEffect_drop_and_pop(durationMs: Float, height: Float): Pointer
    fun AnimationEffect_spin_in(durationMs: Float, turns: Float): Pointer
    fun AnimationEffect_turntable(durationMs: Float): Pointer
    fun AnimationEffect_add_tween(handle: Pointer, propertyName: Slice, from: Float, to: Float, easingName: Slice): ResultUnitInt
    fun AnimationEffect_add_keyframe(handle: Pointer, propertyName: Slice, at: Float, value: Float, easingName: Slice): ResultUnitInt
    fun AnimationEffect_set_repeat_forever(handle: Pointer): Unit
}
/** A reusable set of property tracks, modelled after Anime.js object animations.
*/
class AnimationEffect internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class AnimationEffectCleaner(val handle: Pointer, val lib: AnimationEffectLib) : Runnable {
        override fun run() {
            lib.AnimationEffect_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, AnimationEffect.AnimationEffectCleaner(handle, AnimationEffect.lib));
    }

    companion object {
        internal val libClass: Class<AnimationEffectLib> = AnimationEffectLib::class.java
        internal val lib: AnimationEffectLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        fun create(durationMs: Float): AnimationEffect {
            
            val returnVal = lib.AnimationEffect_create(durationMs);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = AnimationEffect(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun instant(): AnimationEffect {
            
            val returnVal = lib.AnimationEffect_instant();
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = AnimationEffect(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun popIn(durationMs: Float): AnimationEffect {
            
            val returnVal = lib.AnimationEffect_pop_in(durationMs);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = AnimationEffect(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun dropIn(durationMs: Float, height: Float): AnimationEffect {
            
            val returnVal = lib.AnimationEffect_drop_in(durationMs, height);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = AnimationEffect(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun dropAndPop(durationMs: Float, height: Float): AnimationEffect {
            
            val returnVal = lib.AnimationEffect_drop_and_pop(durationMs, height);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = AnimationEffect(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun spinIn(durationMs: Float, turns: Float): AnimationEffect {
            
            val returnVal = lib.AnimationEffect_spin_in(durationMs, turns);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = AnimationEffect(handle, selfEdges, true)
            return returnOpaque
        }
        @JvmStatic
        
        fun turntable(durationMs: Float): AnimationEffect {
            
            val returnVal = lib.AnimationEffect_turntable(durationMs);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = AnimationEffect(handle, selfEdges, true)
            return returnOpaque
        }
    }
    
    /** Add a two-key property tween. Property names follow Anime.js/Three.js:
    *`x`, `y`, `z`, `rotateX`, `rotateY`, `rotateZ`, `scale`, `opacity`,
    *`tintR/G/B/A`, and `emissiveR/G/B`.
    */
    fun addTween(propertyName: String, from: Float, to: Float, easingName: String): Result<Unit> {
        val propertyNameSliceMemory = PrimitiveArrayTools.borrowUtf8(propertyName)
        val easingNameSliceMemory = PrimitiveArrayTools.borrowUtf8(easingName)
        
        val returnVal = lib.AnimationEffect_add_tween(handle, propertyNameSliceMemory.slice, from, to, easingNameSliceMemory.slice);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                return Unit.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            propertyNameSliceMemory.close()
            easingNameSliceMemory.close()
        }
    }
    
    /** Add a normalised keyframe (`at` in `0..=1`) to a property track.
    */
    fun addKeyframe(propertyName: String, at: Float, value: Float, easingName: String): Result<Unit> {
        val propertyNameSliceMemory = PrimitiveArrayTools.borrowUtf8(propertyName)
        val easingNameSliceMemory = PrimitiveArrayTools.borrowUtf8(easingName)
        
        val returnVal = lib.AnimationEffect_add_keyframe(handle, propertyNameSliceMemory.slice, at, value, easingNameSliceMemory.slice);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                return Unit.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            propertyNameSliceMemory.close()
            easingNameSliceMemory.close()
        }
    }
    
    fun setRepeatForever(): Unit {
        
        val returnVal = lib.AnimationEffect_set_repeat_forever(handle);
        
    }

}